### PR TITLE
Code Insights: Add new insight error UI V2

### PR DESCRIPTION
**Pull request by vovakulikov**
Closes https://github.com/sourcegraph/sourcegraph/issues/44961
[Designs ](https://www.figma.com/file/aMO6sPoUHDUUcqa25wFnWl/Surface-timeout-limit-hit-to-end-users-in-the-UI-%2339969-Ready-for-dev?node-id=2034%3A5789&t=maBwfj5IA1hx4O3n-0)

## Background

This PR adds new insight error UI. Previously we had only timeout errors UI and it was pretty limited we didn't show any timeout error UI information (such as list of points that are failed with timeout errors). In this PR we list these points and also we add handling for all other error types (such as Generic error points). See this comment for more details https://github.com/sourcegraph/sourcegraph/issues/44961#issuecomment-1339926367

| Before | After |
| ------- | ------ | 
| Screenshot 2022-12-07 at 02 42 37 | Screenshot 2022-12-07 at 02 42 13 |

## Design review 
- Since I don't see timeout or generic errors on the k8s staging I suggest to use storybook story to check new error UI. For example, in [this storybook demo](https://www.chromatic.com/component?appId=5f0f381c0e50750022dc6bf7&csfId=web-insights-smartinsightsviewgridexample&buildNumber=56072&k=63902e94886b4102a13bb1d3-576-interactive-true&h=4&b=-1)
- @AlicjaSuska Grey error UI icons seemed very bold, In this PR I used icon with --warning (standard color from our palette for warning UI) let me know if we still want to stick with grey one. 
- Ideally I would add hover effect (if you hover point in the error popover we hover this point on the chart) but since we don't have time and this PR already has a lot of changes I suggest iterate on this later. 

## For reviewers 
Please don't look at PR diff size, most of this size is mock changes (I had to change our mocks since I changed the data insight query), Without mocks this PR is standard ~400-450 lines of code. 

## Test plan
- Go to the "Smart Insights View Grid Example" story 
- Check that layout UI next to the legend items looks correct
- Check that layout UI next to the insight title also looks correct 

## App preview:

- [Web](https://sg-web-vk-insight-error-ui.onrender.com/search)

Check out the [client app preview documentation](https://docs.sourcegraph.com/dev/how-to/client_pr_previews) to learn more.

### CODE REVIEW

**Comment by sg-e2e-regression-test-bob** [comment]
## Bundle size report 📦

| Initial size | Total size | Async size | Modules |
| --- | --- | --- | --- |
| 0.00% (+0.12 kb) | 0.02% (+3.52 kb) | 0.03% (+3.40 kb) | 0.00% (0) |

Look at the [Statoscope report](https://console.cloud.google.com/storage/browser/_details/sourcegraph_reports/statoscope-reports/vk/insight-error-ui/compare-report.html;tab=live_object) for a full comparison between the commits 49b3083 and f5e85bb or [learn more](https://docs.sourcegraph.com/dev/how-to/testing#bundlesize).


<details>
<summary>Open explanation</summary>

- `Initial size` is the size of the initial bundle (the one that is loaded when you open the page)
- `Total size` is the size of the initial bundle + all the async loaded chunks
- `Async size` is the size of all the async loaded chunks
- `Modules` is the number of modules in the initial bundle
</details>

**Comment by AlicjaSuska** [comment]
@vovakulikov 

> @AlicjaSuska Grey error UI icons seemed very bold, In this PR I used icon with --warning (standard color from our palette for warning UI) let me know if we still want to stick with grey one.

Let's stick with the grey color (icon-color). Thank you!

File: client/wildcard/src/hooks/useElementObscuredArea.ts [hunk 1]
```diff
@@ -47,13 +48,25 @@ export function useElementObscuredArea<T extends HTMLElement>(
...
         }
-    }, [elementReference, calculate])
+    }, [elementReference, calculate, lazy])
 
     return obscured
 }
+
+function scheduleIntoNextFrame(callback: () => void): void {
+    requestAnimationFrame(() => {
+        requestAnimationFrame(() => {
```
Comment (philipp-spiess) [inline client/wildcard/src/hooks/useElementObscuredArea.ts:68]: Technically that's not the next but the next next frame. It feels a bit arbitrary though. Why is this necessary? Should this be using `requestIdleCallback` instead (at least for browsers that support it)?

**Review by philipp-spiess — APPROVED** [review]
Some comments inline. Overall I’m a BIG FAN of this improvement. Feels a lot more robust with this level of detail when things go wrong. Amazing work!

**Comment by vovakulikov** [comment]
@AlicjaSuska a quick comment on your points
>the message about the error type is displayed below every point, there is no 'affected points' header

I did it because otherwise the current UI doesn't cover all cases, what if we have two timeout error points and 2 different generic errors with custom BE error messages. 

We can put timeout errors in one section within tooltip but I strongly recommend not remove the error messages below the points (In storybook demo they are all the same but in the real world that may be different) [See my Figma comment here](https://www.figma.com/file/aMO6sPoUHDUUcqa25wFnWl?node-id=2034:5789#323572391)

File: client/web/src/enterprise/insights/components/insights-view-grid/components/backend-insight/components/backend-insight-alerts/BackendInsightAlerts.tsx [hunk 2]
```diff
@@ -73,3 +89,107 @@ export const BackendInsightErrorAlert: FC<BackendInsightErrorAlertProps> = props
     ) : (
         <ErrorAlert error={props.error} />
     )
+
+interface InsightIncompleteAlertProps {
+    alert: IncompleteDatapointAlert
+}
+
+export const InsightIncompleteAlert: FC<InsightIncompleteAlertProps> = props => {
```
[the remark this answers is not in the document]
Comment (vovakulikov) [reply client/web/src/enterprise/insights/components/insights-view-grid/components/backend-insight/components/backend-insight-alerts/BackendInsightAlerts.tsx:97]: The main idea was that these components are used at the same time to build and cover all possible cases for error/alert UI for the insight card. I was thinking to have a separate file for these but all three main components here are used in same files. And it also makes sense to group them because these components are from the same family (insight alert UI)

File: client/web/src/enterprise/insights/core/backend/gql-backend/methods/get-backend-insight-data/deserializators.ts [hunk 3]
```diff
@@ -9,24 +9,38 @@ export const createBackendInsightData = (insight: BackendInsight, response: Insi
     const seriesData = response.dataSeries
     const isFetchingHistoricalData = seriesData.some(({ status: { isLoadingData } }) => isLoadingData)
     const isAllSeriesErrored = seriesData.every(series => series.status.incompleteDatapoints.length > 0)
+    const topLevelIncompleteAlert = isAllSeriesErrored ? seriesData[0].status.incompleteDatapoints[0] : null
```
[the remark this answers is not in the document]
Comment (vovakulikov) [reply client/web/src/enterprise/insights/core/backend/gql-backend/methods/get-backend-insight-data/deserializators.ts:12]: Nice catch, I thought that `every` works slightly different and returns false if array is empty but I see that if seriesData is empty it returns true and it will fail on `seriesData[0]`. I will add seriesData length check to the `isAllSeriesErrored` thanks.

File: client/wildcard/src/components/Charts/components/line-chart/components/legend-list/LegendList.module.scss [hunk 4]
```diff
@@ -8,17 +8,31 @@
...
 }
 
+.legend-mark-container {
+    display: flex;
+    align-items: center;
+
+    &::before {
+        // Render an invisible "a" symbol in order to improve non-symbol UI element (legend mark)
+        // alignment with legend text layout
+        content: 'a';
```
[the remark this answers is not in the document]
Comment (vovakulikov) [reply client/wildcard/src/components/Charts/components/line-chart/components/legend-list/LegendList.module.scss:27]: @philipp-spiess Because padding can have only a static value here but we want to have a dynamic value for block with icon (fontSize * lineHeight), So the only one generic way that I could come up with was this solution, render invisible symbol that (by default) has inherited fontSize and lineHeight and this symbol spreads the container by height. 

```html
 <div class="parent">
    <-- a symbol will spread the parent by height -->
   ::before 
    <-- Icon will be placed on the center of the spreader parent block -->
   <span class="icon" /> 
 </div>
```

Classic `display: flex` and `align-items:center` won't work here as we want because we want to center icon only by symbol center of the first text line and not by the whole text block 

| Before | After | 
| ------- | ------- |
| Screenshot 2022-12-07 at 11 36 05 | Screenshot 2022-12-07 at 11 35 47 |

File: client/wildcard/src/hooks/useElementObscuredArea.ts (hunk 1 above)
Comment (vovakulikov) [reply client/wildcard/src/hooks/useElementObscuredArea.ts:68]: It's a terminology problem, by next I mean next after the current frame. Current frame is in `requestAnimationFrame` so the next one right after the current)
which is nested  `requestAnimationFrame`

Safari has scheduled requestAnimationFrame after the current pain so in safari it would be enough to call just one `requestAnimationFrame`. 

The need for this is kind if simple. We want to use ScrollBox UI (scroll container that has these gradient shutters if element has top/bottom scroll). But we want to use ScrollBox inside of popover. Popover itself measures sizes of content and change them in order to fit popover in the viewport. But it brings it back in the current paint (`requestAnimationFrame (() => popoverPosition())`

But this popover measurements breaks ScrollBox initial shutter calculation, the thing that we need to do here is to ensure that the following order is happening 
- Click on the popover trigger
- Render Popover element 
- Render Scrollbox element within popover 
- Popover schedule positions logic in `requestAnimationFrame `
- ScrollBox schedule shutter initiate logic in the next frame `requestAnimationFrame(() => requestAnimationFrame())`
- Popover set popover element sizes
- Scrollbox set shutter if scroll box element has scroll

File: client/wildcard/src/components/Charts/components/line-chart/components/legend-list/LegendList.module.scss (hunk 4 above)
[the remark this answers is not in the document]
Comment (philipp-spiess) [reply client/wildcard/src/components/Charts/components/line-chart/components/legend-list/LegendList.module.scss:27]: Okay, so as far as I understand this is needed to get a height for the container that also houses the icon? I wonder why the `<span class="icon">` does not contribute to the height hm.

Anyway I don't have a better idea of doing this since we can't access the line-height from css (we could get the font size with em but not the line height I think?). If the width doesn't matter maybe a [zero-width space](https://unicode-table.com/en/200B/) could be used instead but otherwise I have no idea 😅

File: client/wildcard/src/hooks/useElementObscuredArea.ts (hunk 1 above)
Comment (philipp-spiess) [reply client/wildcard/src/hooks/useElementObscuredArea.ts:68]: I see. 😢 

Why is the popover doing the repositioning in a different frame though, doesn't this cause visible layout tearing? (i.e we render a _wrong frame_ before we fix it). Is this because the observer APIs that we're using have async callbacks? If that's the case, can we even guarantee that it will be the next frame call? 😅  

I’m asking because it looks like with this logic we'll be spreading re-layouting and re-rendering to three different frames in what ideally could be done in one. I understand that the usage of `useElementObscuredArea` is maybe not super visible on the UI so it might be fine but it's still overhead and it makes it harder to reason about bugs.

File: client/wildcard/src/components/Charts/components/line-chart/components/legend-list/LegendList.module.scss (hunk 4 above)
[the remark this answers is not in the document]
Comment (vovakulikov) [reply client/wildcard/src/components/Charts/components/line-chart/components/legend-list/LegendList.module.scss:27]: > I understand this is needed to get a height for the container that also houses the icon?

This is correct yes.

> I wonder why the <span class="icon"> does not contribute to the height hm.

It does but in this case icon has smaller width/height (smaller than fontSize * lineHeight). Because of this visually this icon is hanging in the air (top border of the text block)

[the remark this answers is not in the document]
Comment (vovakulikov) [reply client/wildcard/src/components/Charts/components/line-chart/components/legend-list/LegendList.module.scss:27]: >we could get the font size with em but not the line height I think?

Yes this is a problem, we have to have a static value somewhere for this, and I didn't like this way this is why I came up with this trick

File: client/wildcard/src/hooks/useElementObscuredArea.ts (hunk 1 above)
Comment (vovakulikov) [reply client/wildcard/src/hooks/useElementObscuredArea.ts:68]: @philipp-spiess 
> Why is the popover doing the repositioning in a different frame 

I think we could do a better logic there you're right, I think I had some problems with sync positioning but I need to revisit it to say it for sure. 

>doesn't this cause visible layout tearing

Nope, since RAF runs before the paint phase and we create tether instance with useLayoutEffect there should be no visual flashing. 


> Is this because the observer APIs that we're using have async callbacks? If that's the case, can we even guarantee that it will be the next frame call?

Can you expand on this I'm not sure I understand the context here

Comment (vovakulikov) [reply client/wildcard/src/hooks/useElementObscuredArea.ts:68]: I think as soon as I remove next frame scheduling in the popover I could delete this double RAF from the `useElementObscuredArea`, but I would like to do it in a follow up PR

Comment (philipp-spiess) [reply client/wildcard/src/hooks/useElementObscuredArea.ts:68]: > , but I would like to do it in a follow up PR

Yeah, absolutel.

**Review by AlicjaSuska — APPROVED** [review]
Thank you for making those changes @vovakulikov. I think it's good to go :)
